src/hotspot/share/runtime/handshake.cpp

```diff
@@ -365,27 +376,30 @@ void HandshakeState::add_operation(HandshakeOperation* op) {
   // Adds are done lock free and so is arming.
   // Calling this method with lock held is considered an error.
   assert(!_lock.owned_by_self(), "Lock should not be held");
-  _queue.add(op);
+  _queue.push(op);
   SafepointMechanism::arm_local_poll_release(_handshakee);
 }
 
 HandshakeOperation* HandshakeState::pop_for_self() {
+  assert(_handshakee == Thread::current(), "Must be called by self");
   assert(_lock.owned_by_self(), "Lock must be held");
   return _queue.pop();
 };
 
-static bool processor_filter(HandshakeOperation* op) {
-  return !op->is_asynch();
+static bool non_self_queue_filter(HandshakeOperation* op) {
```

Comment:
The name "non_self_queue_filter" is really awkward - sorry. But I think we're going to have to revisit the way filtering is named and done once we try to generalise it anyway, so I'll let this pass.